### PR TITLE
systemd: 250.4 -> 251.3

### PR DESCRIPTION
**Pull request by Mic92**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Mic92** [comment]
Patches are up-to-date, but there are some runtime replacements left to do.

File: pkgs/os-specific/linux/systemd/default.nix
Comment (YorikSar) [inline pkgs/os-specific/linux/systemd/default.nix:133]: The releaseTimestamp below should be `1653139508`

File: pkgs/os-specific/linux/systemd/default.nix
Comment (YorikSar) [inline pkgs/os-specific/linux/systemd/default.nix:584]: I guess you meant to add `ignore` support here. I think we shouldn't ignore these files, because `tar` strings in them refer to `systemd-pull tar` argument, which is already replaced with Nix store path in `src/import/pull-tar.c` and others.

Overall `systemd-pull` seems to be broken currently and I think we should fix this in a separate PR.

File: pkgs/os-specific/linux/systemd/0012-inherit-systemd-environment-when-calling-generators.patch
Comment (YorikSar) [inline pkgs/os-specific/linux/systemd/0012-inherit-systemd-environment-when-calling-generators.patch:38]: Here you're not only discarding environment variables that came from outside, but also ones that are generated by `build_generator_environment` function. Our version of the patch prevents that function from copying environment variables from `transient_environment` instead: https://github.com/tweag/nixpkgs/blob/systemd-boot-zfs/pkgs/os-specific/linux/systemd/0012-inherit-systemd-environment-when-calling-generators.patch - this should remove all unneeded variables, while keeping essential ones.

**Review by YorikSar — CHANGES_REQUESTED** [review]
You've beat me to posting this PR. We've already made similar changes in this branch: https://github.com/tweag/nixpkgs/tree/systemd-251/pkgs/os-specific/linux/systemd - maybe you can find something else useful there.

**Comment by catap** [comment]
@Mic92 may I ask you about small favour? Is it possible to wait a bit to jump to v251.3? Why? To include https://github.com/systemd/systemd/pull/23859 which with https://github.com/NixOS/nixpkgs/pull/179582 and this one allows to make detection of parallel guest works well.

File: pkgs/os-specific/linux/systemd/default.nix
Comment (Mic92) [reply pkgs/os-specific/linux/systemd/default.nix:133]: updated.

File: pkgs/os-specific/linux/systemd/0012-inherit-systemd-environment-when-calling-generators.patch
Comment (Mic92) [reply pkgs/os-specific/linux/systemd/0012-inherit-systemd-environment-when-calling-generators.patch]: applied.

**Comment by Mic92** [comment]
> @Mic92 may I ask you about small favour? Is it possible to wait a bit to jump to v251.3? Why? To include [systemd/systemd#23859](https://github.com/systemd/systemd/pull/23859) which with #179582 and this one allows to make detection of parallel guest works out of the box in the installer.

sure. This needs to be tested before anyway.

File: pkgs/os-specific/linux/systemd/default.nix
Comment (Mic92) [inline pkgs/os-specific/linux/systemd/default.nix:175]: @alyssais are there already musl patches for this version of systemd?

**Comment by catap** [comment]
> > @Mic92 may I ask you about small favour? Is it possible to wait a bit to jump to v251.3? Why? To include [systemd/systemd#23859](https://github.com/systemd/systemd/pull/23859) which with #179582 and this one allows to make detection of parallel guest works out of the box in the installer.
> 
> sure. This needs to be tested before anyway.

I'm testing it for a week and it works well. Anyway, I'm installing it via overlay as described inside https://github.com/NixOS/nixpkgs/pull/179582

**Comment by hax404** [comment]
> > > @Mic92 may I ask you about small favour? Is it possible to wait a bit to jump to v251.3? Why? To include [systemd/systemd#23859](https://github.com/systemd/systemd/pull/23859) which with #179582 and this one allows to make detection of parallel guest works out of the box in the installer.
> > 
> > 
> > sure. This needs to be tested before anyway.
> 
> I'm testing it for a week and it works well :) Anyway, I'm installing it via overlay as described inside #179582

Thank you for the hint how to test it. I am using this systemd version to work around https://github.com/systemd/systemd/issues/23197 on NixOS 22.05.

File: pkgs/os-specific/linux/systemd/default.nix
Comment (alyssais) [reply pkgs/os-specific/linux/systemd/default.nix:175]: [Yes](https://git.openembedded.org/openembedded-core/commit/meta/recipes-core/systemd?id=3275cc6c2ecfd70ba134b49363c9cecbb0c2c0b2)

Comment (alyssais) [reply pkgs/os-specific/linux/systemd/default.nix:175]: Usually I'd plan on updating the Musl patches myself after the main systemd PR was merged (to avoid friction, since most of our systemd maintainers don't want to have to think about Musl), but you could also incorporate them into this PR if you'd prefer.

Comment (arianvp) [reply pkgs/os-specific/linux/systemd/default.nix:175]: Happy with either approach. @Mic92 i'm happy to merge as is or wait until you've added the patches

**Comment by catap** [comment]
@Mic92 FYI: 251.3 was released a few days ago.

**Comment by Mic92** [comment]
Bumped to 251.3

**Comment by Mic92** [comment]
Tested on my laptop like this: https://github.com/Mic92/dotfiles/commit/ea348107916f467ce45acceaa2d898f060c8465a

**Comment by vcunat** [comment]
- [ ] this systemd update broke tests of `umockdev`, maintained by @flokli
  https://hydra.nixos.org/build/184809853

**Comment by YorikSar** [comment]
Note that umockdev fixes for system 251: https://github.com/martinpitt/umockdev/pull/181 - landed in [0.17.10](https://github.com/martinpitt/umockdev/releases/tag/0.17.10), we have 0.17.9 at this moment: https://github.com/NixOS/nixpkgs/blob/a1439bc96590dae913369f4515ca3df6ca1a1556/pkgs/development/libraries/umockdev/default.nix#L22.

It seems we need to just bump umockdev.

**Comment by arianvp** [comment]
Please open a PR. I happily merge ir

**Comment by vcunat** [comment]
Just note, the fixes should be targeted at the `staging-next` branch.

**Comment by YorikSar** [comment]
Posted PR https://github.com/NixOS/nixpkgs/pull/182668, please check it out.

File: pkgs/os-specific/linux/systemd/default.nix
Comment (vcunat) [reply pkgs/os-specific/linux/systemd/default.nix:175]: @alyssais: it's in master, and musl is failing.  It doesn't suffice to simply drop the subset that doesn't apply.

Comment (alyssais) [reply pkgs/os-specific/linux/systemd/default.nix:175]: Whoops, I missed that this was merged.  I'll get that fixed.

Comment (vcunat) [reply pkgs/os-specific/linux/systemd/default.nix:175]: Cross-ref: https://github.com/NixOS/nixpkgs/pull/183721

**Pull request merged**
